### PR TITLE
Bluetooth

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

https://expo.canny.io/feature-requests/p/bluetooth-api

# How

Not very committed to this API at the moment. iOS only right now.

Currently the listener system is very incomplete and needs some kind of tagging system to get updates regarding input events.

* `startScanAsync()`
* `stopScanAsync()`
* `connectAsync()`
* `disconnectAsync()`
* `discoverServicesAsync()`
* `readAsync()`
* `writeAsync()`
* `updateAsync()`
* `readRSSIAsync()`
* `getPeripheralsAsync()`
* `getCentralAsync()`
* `isScanningAsync()`
* `addListener()`
* `removeSubscription()`
* `removeAllListeners()`

# Test Plan

Added a screen to NCL that lets you poll for devices, connect/disconnect to any device, finally read or write to the device.

### CODE REVIEW

**Comment by EvanBacon** [comment]
Currently you can do an operation like:
```js

Bluetooth.addListener(({event, data}) => {
  if (event === Bluetooth.Events.CENTRAL_DID_CONNECT_PERIPHERAL_EVENT) {
    // Peripheral connected
  } else if (event === Bluetooth.Events.CENTRAL_DID_FAIL_TO_CONNECT_PERIPHERAL_EVENT) {
    // Peripheral failed (should throw error?)
  }
})
Bluetooth.connectAsync(...);

```

But ideally you would do

```js
try {
  const data = await Bluetooth.connectAsync(...);
} catch (error) {
}
```

File: packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth.m [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+{
+  for (CBService *service in peripheral.services) {
+    if ([service.UUID.UUIDString isEqualToString:UUID.UUIDString]) {
+      return service;
+    }
+  }
+  return nil;
+}
+
+- (CBCharacteristic *)characteristicFromUUID:(CBUUID *)UUID service:(CBService *)service prop:(CBCharacteristicProperties)prop
```
Comment (ide) [inline packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth.m]: Write out:

```suggestion
- (CBCharacteristic *)characteristicFromUUID:(CBUUID *)UUID service:(CBService *)service properties:(CBCharacteristicProperties)properties
```

I think it makes sense to document (or rename) whether this method is supposed to return a characteristic with any of the props or all of the props.

File: packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth.m [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+      
+      NSData *data = [[NSData alloc] initWithBase64EncodedString:options[@"data"] options:NSDataBase64DecodingIgnoreUnknownCharacters];
+      if (!data) {
+        reject(@"E_INVALID_FORMAT", @"Failed to parse base64 string.", nil);
+        return;
+      }
+      [peripheral writeValue:data forCharacteristic:characteristic type:writeType];
+    }
+  }
+  resolve([NSNull null]);
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth.m]: @tsapeta  suggested `[NSNull null]` for resolves a little while back. Is there a reason to use `nil` instead?

**Comment by watadarkstar** [comment]
Great to see this as a PR! Looking forward to seeing this progress! 🎉

File: packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth+JSON.m [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+}
+
++ (NSMutableArray<CBUUID *> *)CBUUIDList_JSONToNative:(NSArray *)input
+{
+  if (!input) return nil;
+  
+  NSMutableArray *output = [NSMutableArray new];
+  for (id value in input) {
+    if ([value isKindOfClass:[NSString class]]) {
+      // TODO: Bacon: Maybe add support for generating CBUUID from Data (ArrayBuffer)
```
Comment (ide) [inline packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth+JSON.m]: Probably should stick to strings, we're doing BLE I/O here, using strings vs. ArrayBuffers is not a meaningful performance comparison and in terms of an API IMO strings are more natural to work with

File: packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth+JSON.m [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  for (CBUUID *value in input) {
+    NSString *serializedValue = [[self class] CBUUID_NativeToJSON:value];
+    [output addObject:EXNullIfEmpty(serializedValue)];
+  }
+  return output;
+}
+
++ (NSMutableArray<CBUUID *> *)CBUUIDList_JSONToNative:(NSArray *)input
+{
+  if (!input) return nil;
```
Comment (ide) [inline packages/expo-bluetooth/ios/EXBluetooth/EXBluetooth+JSON.m]: Should check that we don't get NSNull passed in as well..

**Comment by serenecloud** [comment]
@EvanBacon watching this PR come along is getting us very excited in the office. Thank you for your efforts!

**Comment by giautm** [comment]
Is this module will work with BLE? Like Beacon (https://www.beaconstac.com/what-is-a-bluetooth-beacon)?

**Comment by EvanBacon** [comment]
@giautm this library will use Bluetooth Low Energy, but that is unrelated to beacons.

**Comment by gre** [comment]
@EvanBacon is there a way to test this on a react-native install? (with unimodules?)

**Comment by EvanBacon** [comment]
@gre I haven't updated it with the latest. Cloning the monorepo and working in `apps/` is probably the easiest approach right now.

File: android/expoview/build.gradle [hunk 5]
```diff
@@ -208,6 +208,7 @@ dependencies {
   api project(':expo-background-fetch')
   api project(':expo-barcode-scanner')
   api project(':expo-barcode-scanner-interface')
+  api project(':expo-bluetooth')
```
Comment (EvanBacon) [inline android/expoview/build.gradle]: @ide should we remove expo-bluetooth from expo client until it's more stable? Seeing as it adds another permission we may not be able to add it anyways.

**Comment by EvanBacon** [comment]
Moving some inline comments to the PR:

## Native Component List Demo

- Mounting this screen should start scanning for devices right away.
  - This should collect all of the available bluetooth devices.
  - This should filter out various bluetooth devices that aren't helpful. TODO: Bacon: (No name?)
  - Devices should be sorted by discovery time to prevent jumping around.
- Display them in a SectionList with one section dedicated to connected devices, and the other for every other device
- Each device should have a button to view more info.
  - The more info screen should contain a button to disconnect if connected.
  - A button to (Forget This Device). Is this possible?
  - The ability to rewrite data - example airpods allow for a name change
  - A list of known functionality.
- If I tap a non-connected device I should be able to attempt a connection.
  - If the connection fails (or times-out?) there should be an alert: "Connection Unsuccessful" "Make sure 'name of device' is turned on and in range." "OK"
- When a device is connecting it should have an indicator next to it.
- Is it possible to get a list of known devices like in the iOS bluetooth settings?
- There should be an indicator at the bottom of the list which shows that scanning is in progress
- Is it possible to show the discoverable name of the current device (ios settings)
- There should be a toast that presents non-intrusive errors like a device disconnecting from us.
- There should be an alert for larger errors.

- There should be a section dedicated to the manager.
- This section will display the state of the manager.
- There should be a toggle for scanning.
- If possible (on iOS) we should link to Bluetooth in settings.
- On Android we should be able to turn on/off bluetooth

### Extra:
- Observe when a device disconnects
- Search by name

## Use-cases

### Get location of parked car

From `@andrioid`

This won't require bluetooth, I should create an example and document this to cut down on questions.

### Preserve connection across hot reloads

From Canny

When a developer is testing bluetooth, they may want to keep the bluetooth state. This could however create leaks and inconsistent state design.

### I just want to scan for peripherals, I'll update my UI from within the view.

```ts
Bluetooth.startScanningAsync({}, ({ peripheral }) => {
  // Update the view state
  this.setState(({ peripherals }) => {
    const { [peripheral.id]: currentPeripheral = {}, ...others } = peripherals;
    return {
      peripherals: {
        ...others,
        [peripheral.id]: peripheral
      },
    };
  });
});
```

## Resources:

* List of service UUIDs: https://www.bluetooth.com/specifications/gatt/services

File: ios/Exponent/Supporting/Info.plist [hunk 6]
```diff
@@ -113,6 +113,8 @@
 	<true/>
 	<key>UIBackgroundModes</key>
 	<array>
+		<string>bluetooth-central</string>
```
Comment (EvanBacon) [inline ios/Exponent/Supporting/Info.plist]: Should remove these if we aren't supporting the feature

File: android/expoview/build.gradle (hunk 5 above)
Comment (ide) [reply android/expoview/build.gradle]: Yes, let's leave it out and not touch permissions.

File: ios/Exponent/Supporting/Info.plist (hunk 6 above)
Comment (ide) [reply ios/Exponent/Supporting/Info.plist]: Let's remove these for now, and instead have a guide with the unimodule explaining what plist keys need to be set.

File: packages/expo-bluetooth/README.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+## Introduction
+
+Expo Bluetooth provides access to native Bluetooth Low-Energy features in React Native.
+
+### What's not included
+
+- Background scanning: Because of how strict Apple is about this permission, I've prioritized it very low.
+- Restoration: React Native state mixed with the asynchronous nature of the native Bluetooth API makes this a very messy and unreliable feature.
+- Mesh Networks: Basically completely unrealted to this library.
```
Comment (vpontis) [inline packages/expo-bluetooth/README.md]:
```suggestion
- Mesh Networks: Basically completely unrelated to this library.
```

**Comment by OmniAsetty** [comment]
Thank you so much for working on this! Is this feature getting added in Expo v33? We are going to start working on Bluetooth app in June first week. Do you think this feature will be ready by then?

**Comment by zocheyado** [comment]
Is this feature going include bluetooth serial or just bluetooth low energy? Planning a future project but I suspect this will only be low energy.

**Comment by JoseCarlosNS** [comment]
I am working on a project now and the abscence of bluetooth support is the ONLY reason keeping me from using Expo to build my apps, for bluetooth support is essential in my app.

Also, will this feature support bluetooth serial? I hope it does.

**Comment by ide** [comment]
This module is only for a subset of BLE, namely scanning for peripherals. Broadcasting as a peripheral is out of scope for this module. Additionally this module will be part of the bare workflow due to extra permissions needed to use Bluetooth. The native Bluetooth libraries are inconsistent/broken enough that doing a quality job on a subset of the BLE APIs is a big investment.

We plan not to work on Bluetooth serial due to the fact that it is complex to maintain despite the fact that few apps need it, and on iOS you need to enroll in MFi.

To use Expo with Bluetooth serial you would use the bare workflow and then write your own Bluetooth serial module.

**Comment by aquaspirit26** [comment]
how is this project going? we are a naive team of embedded firmware and cloud engineers, very curious if this is going well-enough to follow. support from us would be happily provided.

**Comment by diogoviannaaraujo** [comment]
I gave one more on that canny feature request about 3 years ago...

**Comment by TrustyTechSG** [comment]
Any update on this? Bluetooth is the only thing stop me from using Expo.

**Comment by ozerty** [comment]
Hi guys, what happens with this PR ? It will be a nice feature for both IOS and Android.
Thanks

**Comment by ide** [comment]
For the Expo platform (the client code) we're mostly investing in making it so developers who have bespoke needs can write modules for their own use cases and so the Expo team is not a bottleneck. On top of that, the scope of Bluetooth is very broad and asking for Bluetooth is kind of like asking for "internet" which can mean a lot of things. For the time being if you need some subset of Bluetooth functionality in your Expo app the best way to achieve that is to use the bare workflow from which point it's mostly straightforward Obj-C and Java for your needs at hand.

**Comment by ozerty** [comment]
Got it ! Thanks for the fast reply.

**Pull request closed without merging**